service/gcs/core/gcs/gcs.go

```diff
@@ -542,32 +558,60 @@ func (c *gcsCore) ResizeConsole(pid int, height, width uint16) error {
 	return p.Tty.ResizeConsole(height, width)
 }
 
-// WaitContainer waits for a container to complete and returns its exit code.
-func (c *gcsCore) WaitContainer(id string) (int, error) {
+// WaitContainer returns a function that can be used to sucessfully wait for a
+// container exit code. This will only return after all writers on WaitProcess
+// have completed. On error the container id was not a valid container.
+func (c *gcsCore) WaitContainer(id string) (func() int, error) {
 	c.containerCacheMutex.Lock()
 	entry := c.getContainer(id)
 	if entry == nil {
 		c.containerCacheMutex.Unlock()
-		return -1, errors.WithStack(gcserr.NewContainerDoesNotExistError(id))
+		return nil, errors.WithStack(gcserr.NewContainerDoesNotExistError(id))
 	}
 	c.containerCacheMutex.Unlock()
 
-	entry.exitWg.Wait()
-	return entry.exitCode, nil
+	f := func() int {
+		entry.initProcess.writersWg.Wait()
+		return entry.initProcess.exitCode
+	}
+
+	return f, nil
 }
 
-// WaitProcess waits for a process to complete and returns its exit code.
-func (c *gcsCore) WaitProcess(pid int) (int, error) {
+// WaitProcess returns a function that can be used to successfully wait for a
+// process exit code. On error the pid was not a valid pid.
+func (c *gcsCore) WaitProcess(pid int) (func() int, error) {
 	c.processCacheMutex.Lock()
 	entry, ok := c.processCache[pid]
 	if !ok {
 		c.processCacheMutex.Unlock()
-		return -1, errors.WithStack(gcserr.NewProcessDoesNotExistError(pid))
+		return nil, errors.WithStack(gcserr.NewProcessDoesNotExistError(pid))
 	}
 	c.processCacheMutex.Unlock()
 
-	entry.exitWg.Wait()
-	return entry.exitCode, nil
+	// Safe to do without lock because this is a write once variable at creation.
+	if entry.isInitProcess {
+		// We only `Add` to the writers lock for the 2nd..Nth waiters because we
+		// added the 1st writer at creation to ensure that the container exited
+		// notification wasn't written until at least one `WaitProcess` was called.
+		entry.syncRoot.Lock()
+		if entry.writersCalled {
+			entry.writersWg.Add(1)
+		} else {
+			entry.writersCalled = true
+		}
+		entry.syncRoot.Unlock()
+	}
+
+	f := func() int {
+		entry.exitWg.Wait()
+		if entry.isInitProcess {
+			entry.writersWg.Done()
```

Comment:
Is it safe to set this done before the write has actually been made?

Reply:
I agree its a race. I'm not sure how I could possibly do this. I would need to expose a "written" response channel from the bridge.Write and we would have to only release the `writersWg` when we received that message. And even that would not be *truly* enough. Did the socket actually write? Its almost like we would need a two-phase with the HCS to know. IE: Write -> HCS -> Written. Ok now release. Thoughts?

Reply:
Actually... I was wrong. The bridge serializes messages once the call to "Write" has occurred so I can do this. Let me think about how to fix it.